### PR TITLE
[PRODDEV-90] Added files for migrate content for Zoom API

### PR DESCRIPTION
**Pull request by szherebchuk**

### CODE REVIEW

File: src/Client/ZoomSyncClient.php [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+namespace Drupal\zoom_sync\Client;
+
+use Drupal\Component\Serialization\Json;
+use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\zoom_sync\ZoomSyncClientInterface;
+use GuzzleHttp\ClientInterface;
+use GuzzleHttp\Exception\RequestException;
+use Psr\Log\LoggerInterface;
+
```
Comment (anpolimus) [inline src/Client/ZoomSyncClient.php:11]: coding standards

File: src/Client/ZoomSyncClient.php [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+   * @param array $params
+   *
+   * @return mixed
+   */
+  public function get($url, array $params = []) {
+    return $this->request('GET', $url, $params);
+  }
+
+  /**
+   * @return array
```
Comment (anpolimus) [inline src/Client/ZoomSyncClient.php]: Detailed description for the method

File: src/Client/ZoomSyncClient.php [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+   */
+  private function getUsersList() {
+    $start_page = 1;
+    $users_list = [];
+    $users = $this->get('/users');
+    $page_count = $users['page_count'];
+    $page_number = $users['page_number'];
+
+    while ($page_number < $page_count + 1) {
```
Comment (anpolimus) [inline src/Client/ZoomSyncClient.php]: Basic for cycle would be more readable here.

File: src/Client/ZoomSyncClient.php [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+      $result = $this->get('/users', ['page_number' => $start_page]);
+      $users_list = array_merge($users_list, $result['users']);
+      $page_number++;
+    }
+
+    return $users_list;
+  }
+
+  /**
+   * @param $email
```
Comment (anpolimus) [inline src/Client/ZoomSyncClient.php]: Description

File: src/Client/ZoomSyncClient.php [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+   * @param string $type
+   *
+   * @return mixed
+   */
+  private function getMeetings($email, string $type = 'upcoming') {
+    return $this->get( '/users/' . $email . '/meetings', ['type' => $type]);
+  }
+
+  /**
+   * @param $meeting_id
```
Comment (anpolimus) [inline src/Client/ZoomSyncClient.php]: Description

File: src/ZoomSyncClientInterface.php [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+   * @param array $query
+   *   Any Query Parameters defined in the API spec.
+   * @param array $body
+   *   Array that will get converted to JSON for some requests.
+   *
+   * @return object
+   *   \GuzzleHttp\Psr7\Response body
+   */
+  public function request(string $method, string $endpoint, array $query = [], array $body = []);
+
```
Comment (anpolimus) [inline src/ZoomSyncClientInterface.php:29]: descr

File: src/ZoomSyncClientInterface.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  public function get($url, array $params = []);
+
```
Comment (anpolimus) [inline src/ZoomSyncClientInterface.php:38]: descr

File: zoom_sync.info.yml [hunk 8]
```diff
@@ -0,0 +1,9 @@
+name: Zoom Sync
+type: module
+description: 'Provides functionality interacting with the Zoom API.'
+package: Custom
+core: 8.x
```
Comment (anpolimus) [inline zoom_sync.info.yml]:
```suggestion
```

File: zoom_sync.info.yml [hunk 9]
```diff
@@ -0,0 +1,9 @@
...
+core_version_requirement: ^8 || ^9
+configure: zoom_sync.settings
+dependencies:
+  - zoomapi
```
Comment (anpolimus) [inline zoom_sync.info.yml]:
```suggestion
```

File: zoom_sync.info.yml [hunk 10]
```diff
@@ -0,0 +1,9 @@
+name: Zoom Sync
+type: module
+description: 'Provides functionality interacting with the Zoom API.'
```
Comment (anpolimus) [inline zoom_sync.info.yml]:
```suggestion
description: 'Zoom API sync module for Virtual Y'
```

File: zoom_sync.info.yml [hunk 11]
```diff
@@ -0,0 +1,9 @@
...
+package: Custom
```
Comment (anpolimus) [inline zoom_sync.info.yml]:
```suggestion
package: Open Y Virtual YMCA
```

File: migrations/migrate_plus.migration.virtual_meetings.yml [hunk 12]
```diff
@@ -0,0 +1,157 @@
+langcode: en
+status: true
+dependencies:
+  enforced:
+    module:
+      - zoom_sync
+id: virtual_meeting_synchronization
+migration_tags: {}
+migration_group: zoom_sync
+label: 'Create virtual meeting from Zoom API'
```
Comment (anpolimus) [inline migrations/migrate_plus.migration.virtual_meetings.yml]:
```suggestion
label: 'Zoom events sync to Virtual Y'
```

File: src/Plugin/migrate/source/ZoomSyncMeetingSource.php [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+
+        if (count($occurrences) > 1) {
+          $end_time = str_replace('Z', '', end($occurrences)['start_time']);
+        } else {
+          $end_time = strtotime($start_time) + $duration * 60;
+        }
+
+        $recurring_date = [
+          'value' => date('Y-m-d\TH:i:s', $start_time),
+          'end_value' => date('Y-m-d\TH:i:s', $end_time),
```
Comment (anpolimus) [inline src/Plugin/migrate/source/ZoomSyncMeetingSource.php]: Let's try to avoid date transformation.

**Review by anpolimus — CHANGES_REQUESTED** [review]

File: openy_gc_zoom_sync.info.yml [hunk 14]
```diff
@@ -1,9 +1,7 @@
-name: Zoom Sync
+name: OpenY Zoom Sync
 type: module
 description: 'Provides functionality interacting with the Zoom API.'
 package: Custom
 core: 8.x
```
Comment (anpolimus) [inline openy_gc_zoom_sync.info.yml]:
```suggestion
```

**Review by anpolimus — CHANGES_REQUESTED** [review]

File: openy_gc_zoom_sync.routing.yml [hunk 15]
```diff
@@ -0,0 +1,7 @@
+openy_gc_zoom_sync.settings:
+  path: '/admin/openy/integrations/openy_gc_zoom_sync/settings'
+  defaults:
+    _title: 'Zoom Sync Settings'
```
Comment (anpolimus) [inline openy_gc_zoom_sync.routing.yml]:
```suggestion
    _title: 'Virtual Y Zoom Sync Settings'
```

File: openy_gc_zoom_sync.routing.yml [hunk 16]
```diff
@@ -0,0 +1,7 @@
...
+    _form: 'Drupal\openy_gc_zoom_sync\Form\ZoomSyncSettingsForm'
+  requirements:
+    _permission: 'administer configuration'
```
Comment (anpolimus) [inline openy_gc_zoom_sync.routing.yml]:
```suggestion
    _permission: 'administer gated content configuration'
```

File: src/Plugin/migrate/process/ZoomSyncDateProcess.php [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+ *       plugin: zoom_sync_date_proccess
+ *       source: source_value
+ * @endcode
+ *
+ * @MigrateProcessPlugin(
+ *   id = "zoom_sync_date_proccess"
+ * )
+ */
+class ZoomSyncDateProcess extends ProcessPluginBase {
+  public function transform($value, MigrateExecutableInterface $migrate_executable, Row $row, $destination_property) {
```
Comment (anpolimus) [inline src/Plugin/migrate/process/ZoomSyncDateProcess.php:31]: CS

File: src/Plugin/migrate/source/ZoomSyncCategorySource.php [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+    ];
+  }
+
+  /**
+   * @return \ArrayIterator
+   */
+  protected function initializeIterator() {
+    $categories = $this->zoomSyncClient->getCategories();
+
+    $data[0] = [
```
Comment (anpolimus) [inline src/Plugin/migrate/source/ZoomSyncCategorySource.php]: Please, delete

File: src/Plugin/migrate/source/ZoomSyncInstructorSource.php [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+    ];
+  }
+
+  /**
+   * @return \ArrayIterator
+   */
+  protected function initializeIterator() {
+    $instructors = $this->zoomSyncClient->getInstructors();
+
+    $data[0] = [
```
Comment (anpolimus) [inline src/Plugin/migrate/source/ZoomSyncInstructorSource.php]: Please, delete

File: src/Plugin/migrate/source/ZoomSyncMeetingSource.php [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+      $r_type = '';
+      $recurrence = $meeting['recurrence'] ?? NULL;
+      $occurrences = $meeting['occurrences'] ?? NULL;
+
+      if (isset($occurrences)) {
+        $start_time = str_replace('Z', '', $occurrences[0]['start_time']);
+        $duration = $occurrences[0]['duration'];
+
+        if (count($occurrences) > 1) {
+          $end_time = date('Y-m-d\TH:i:s', strtotime(end($occurrences)['start_time']));
```
Comment (anpolimus) [inline src/Plugin/migrate/source/ZoomSyncMeetingSource.php]: Please, use default Drupal aproach for the date.

File: src/Client/ZoomSyncClient.php [hunk 21]
```diff
@@ -0,0 +1,297 @@
...
+   * @return mixed
+   *   RequestException or \GuzzleHttp\Psr7\Response body
+   */
+  public function request(string $method, string $endpoint, array $query = [], array $body = []) {
+    try {
+      $response = $this->httpClient->{$method}(
+        $this->baseUri . $endpoint,
+        $this->buildOptions($query, $body)
+      );
+      // TODO: Add additional response options.
```
Comment (podarok) [inline src/Client/ZoomSyncClient.php:94]: Is this added to Jira?

Comment (podarok) [reply src/Client/ZoomSyncClient.php:94]: Let's ensure this is added to Release notes at least, please

File: src/Client/ZoomSyncClient.php [hunk 22]
```diff
@@ -0,0 +1,297 @@
...
+      $payload = Json::decode($response->getBody()->getContents());
+      return $payload;
+    }
+    catch (RequestException $exception) {
+      // Log Any exceptions.
+      $this->logger->error('Failed to complete Zoom API Task "%error"', [
```
Comment (podarok) [inline src/Client/ZoomSyncClient.php:100]: Let's answer to "so what?" question with action item for the user
Like - please contact support by calling: 123456678, or - "check Zoom settings at "link to form"

File: src/Form/ZoomSyncSettingsForm.php [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+  protected function getEditableConfigNames() {
+    return ['openy_gc_zoom_sync.settings'];
+  }
+
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    $config = $this->config('openy_gc_zoom_syncs.settings');
+
+    $form['help'] = [
+      '#type' => 'markup',
+      '#markup' => $this->t('This configuration form provides settings of Zoom meetings integration.'),
```
Comment (podarok) [inline src/Form/ZoomSyncSettingsForm.php:31]: Can we have a link to Zoom docs here in description where to get URL and Token for Data Service URL?

**Review by podarok — CHANGES_REQUESTED** [review]
Small changes, please

File: src/Client/ZoomSyncClient.php (hunk 21 above)
Comment (anpolimus) [reply src/Client/ZoomSyncClient.php:94]: @podarok I'll do readme for this project separately

**Review by anpolimus — APPROVED** [review]
@podarok I want to merge this PR but leave your comments and my comments for the next small PR.
Our goal is to prepare it for QA today. Our QA will check the content and Serhii will do final fixes in parallel.

**Comment by podarok** [comment]
> @podarok I want to merge this PR but leave your comments and my comments for the next small PR. Our goal is to prepare it for QA today. Our QA will check the content and Serhii will do final fixes in parallel.

Sure, go ahead
Let's ensure we haven't lost these and fix them in follow up

**Pull request merged**